pkgs/os-specific/linux/nvidia-x11/generic.nix

```diff
@@ -0,0 +1,81 @@
+{ version
+, sha256_32bit
+, sha256_64bit
+, settingsSha256
+, persistencedSha256
+, useGLVND ? true
+, preferGtk2 ? false
+}:
+
+{ stdenv, callPackage, callPackage_i686, buildEnv, fetchurl
+, kernel ? null, xorg, zlib, perl, nukeReferences
+, # Whether to build the libraries only (i.e. not the kernel module or
+  # nvidia-settings).  Used to support 32-bit binaries on 64-bit
+  # Linux.
+  libsOnly ? false
+}:
+
+with stdenv.lib;
+
+assert (!libsOnly) -> kernel != null;
+
+let
+  nameSuffix = optionalString (!libsOnly) "-${kernel.version}";
+  pkgSuffix = optionalString (versionOlder version "304") "-pkg0";
+
+  self = stdenv.mkDerivation {
+    name = "nvidia-x11-${version}${nameSuffix}";
+
+    builder = ./builder.sh;
+
+    src =
+      if stdenv.system == "i686-linux" then
+        fetchurl {
+          url = "http://download.nvidia.com/XFree86/Linux-x86/${version}/NVIDIA-Linux-x86-${version}${pkgSuffix}.run";
+          sha256 = sha256_32bit;
+        }
+      else if stdenv.system == "x86_64-linux" then
+        fetchurl {
+          url = "http://download.nvidia.com/XFree86/Linux-x86_64/${version}/NVIDIA-Linux-x86_64-${version}${pkgSuffix}.run";
+          sha256 = sha256_64bit;
+        }
+      else throw "nvidia-x11 does not support platform ${stdenv.system}";
+
+    inherit version useGLVND;
+    inherit (stdenv) system;
+
+    outputs = [ "out" ] ++ optional (!libsOnly) "bin";
```

Comment:
The convention has been to put the output containing binaries as the first output (it changed at one point).

Also I didn't check too closely, but doesn't splitting to multiple outputs remove the need for all the `libsOnly` logic?